_posts/2018-09-28-SpringOne-fun-with-the-functional-web-framework.md

```diff
@@ -0,0 +1,227 @@
+In Web MVC the return type could be a lot of different things like any type of Object, a ResponseEntity, etc.
+
+
+### The RouterFunction
+Is a function that takes a ServerRequest and returns a HandlerFunction using a RequestPredicate.
+
+``` java 
+@Bean
+public RouterFunction<ServerResponse> routerFunction(PetHandler petHandler) {
+    RouterFunction<ServerResponse> html = route()
+            .GET("/pets/{id}", accept(TEXT_HTML), petHandler::renderPet)
+            .GET("/pets", accept(TEXT_HTML), petHandler::renderPets)
+            .build();
+
+    RouterFunction<ServerResponse> json = route()
+            .GET("/pets/{id}", accept(APPLICATION_JSON), petHandler::showPet)
+            .GET("/pets", accept(APPLICATION_JSON), petHandler::showPets)
+            .build();
+
+    return html.and(json);
+}
+```
+
+The order in which you define these router functions matters.
+The first router function's  handler that matches your HTTP request will be the one that is executed.
+This makes it a lot clearer when you read the router functions to know which one will be executed, it's the first one that you define and matches.
+
+An advantage of the RouterFunction over the annotational style is that you can map multiple endpoints to the same HandlerFunction.
+This is not possible in Web MVC becuase you can only put one @RequestMapping on a controller method.
+In the WebFlux.fn framework however, you can refer to one HandlerFunction in as many RouterFunction matchers as you want.
+
+#### Improvements in the RouterFunction spring framework 5.1:
+
+1. A router DSL with less static imports:
+    ``` java
+    //5.0 version
+    route(GET("/people"), personHandler::getPeople)
+    //5.1 version
+    route()
+      .GET("/people"), personHandler::getPeople)
+    ```
+2. And a new pattern matcher to resolve which HandlerFunction to call, which is a lot faster than the previous one.
+
+#### RequestPredicates
+Is a function that maps a ServerRequest to a boolean.
+
+This is used to match your HandlerFunction to a HTTP request.
+
+Spring provides a lot of default predicates for paths, accept headers, etc.
+But you can also create your own very easily, with lambdas, methods, or classes.
+
+``` java
+// lambda
+route().GET("/people", serverRequest -> serverRequest.path().endsWith(".json"), personHandler::getPeople)
+
+
+// method
+route().GET("/people", this::pathEndsWithJson, personHandler::getPeople)
+private boolean pathEndsWithJson(ServerRequest request) {
+    return request.path().endsWith(".json");
+}
+
+// class
+route().GET("/people", new PathEndsWithJsonPredicate(), personHandler::getPeople)
+public class PathEndsWithJsonPredicate implements RequestPredicate {
+    @Override
+    public boolean test(final ServerRequest request) {
+        return request.path().endsWith(".json");
+    }
+}
+```
+
+#### nested RouterFunction
+which is similar to the class level @RequestMapping , but a lot more powerful.
```

Comment:
Would rewrite this slightly:
Similar to the class level `@RequestMapping`, but a lot more powerful. (also note that there was an extra space before the comma)